SimulationAgent/SimulationThreads/DeviceReplayTask.cs

```diff
@@ -0,0 +1,74 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Concurrency;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Diagnostics;
+
+namespace Microsoft.Azure.IoTSolutions.DeviceSimulation.SimulationAgent.SimulationThreads
+{
+    public interface IDeviceReplayTask
+    {
+        Task RunAsync(
+            ConcurrentDictionary<string, ISimulationManager> simulationManagers,
+            CancellationToken runningToken
+        );
+    }
+
+    public class DeviceReplayTask : IDeviceReplayTask
+    {
+        // Global settings, not affected by hub SKU or simulation settings
+        private readonly IAppConcurrencyConfig appConcurrencyConfig;
+
+        private readonly ILogger log;
+
+        public DeviceReplayTask(
+            IAppConcurrencyConfig appConcurrencyConfig,
+            ILogger logger)
+        {
+            this.appConcurrencyConfig = appConcurrencyConfig;
+            this.log = logger;
+        }
+
+        public async Task RunAsync(
+            ConcurrentDictionary<string, ISimulationManager> simulationManagers,
+            CancellationToken runningToken)
+        {
+            var tasks = new List<Task>();
+
+            while (!runningToken.IsCancellationRequested)
+            {
+                // TODO: Do work
+                // foreach (var manager in simulationManagers)
+                // {
+                // }
+
+                var before = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
+
+                // Wait for any pending tasks.
+                if (tasks.Count > 0)
+                {
+                    await Task.WhenAll(tasks);
+                    tasks.Clear();
+                }
+
+                var durationMsecs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - before;
+                this.log.Debug("Device-state loop completed", () => new { durationMsecs });
```

Comment:
I believe the log sentence should not be about device state, I guess this was copied from another file?

Reply:
oops, yep it was copied